google/cloud/aiplatform/featurestore/entity_type.py

```diff
@@ -381,3 +386,836 @@ def delete_features(self, feature_ids: List[str], sync: bool = True,) -> None:
+                f"get {type(data_source)} instead. "
+            )
+
+        if isinstance(feature_time, str):
+            import_feature_values_request.feature_time_field = feature_time
+        elif isinstance(feature_time, datetime.datetime):
+            import_feature_values_request.feature_time = utils.get_timestamp_proto(
+                time=feature_time
+            )
+        else:
+            raise ValueError(
+                f"The type of `feature_time` field should be: `str` or `datetime.datetime`, "
+                f"get {type(feature_time)} instead. "
+            )
+
+        if entity_id_field is not None:
+            import_feature_values_request.entity_id_field = entity_id_field
+
+        if disable_online_serving is not None:
+            import_feature_values_request.disable_online_serving = (
+                disable_online_serving
+            )
+
+        if worker_count is not None:
+            import_feature_values_request.worker_count = worker_count
+
+        return import_feature_values_request
+
+    @base.optional_sync(return_input_arg="self")
+    def _import_feature_values(
+        self,
+        import_feature_values_request: gca_featurestore_service.ImportFeatureValuesRequest,
+        request_metadata: Optional[Sequence[Tuple[str, str]]] = (),
+        sync: bool = True,
+    ) -> "EntityType":
+        """Imports Feature values into the Featurestore from a source storage.
+
+        Args:
+            import_feature_values_request (gca_featurestore_service.ImportFeatureValuesRequest):
+                Required. Request message for importing feature values.
+            request_metadata (Sequence[Tuple[str, str]]):
+                Optional. Strings which should be sent along with the request as metadata.
+            sync (bool):
+                Optional. Whether to execute this import synchronously. If False, this method
+                will be executed in concurrent Future and any downstream object will
+                be immediately returned and synced when the Future has completed.
+
+        Returns:
+            EntityType - The entityType resource object with imported feature values.
+        """
+        _LOGGER.log_action_start_against_resource(
+            "Importing", "feature values", self,
+        )
+
+        import_lro = self.api_client.import_feature_values(
+            request=import_feature_values_request, metadata=request_metadata,
+        )
+
+        _LOGGER.log_action_started_against_resource_with_lro(
+            "Import", "feature values", self.__class__, import_lro
+        )
+
+        import_lro.result()
+
+        _LOGGER.log_action_completed_against_resource(
+            "feature values", "imported", self
+        )
+
+        return self
+
+    def ingest_from_bq(
+        self,
+        feature_ids: List[str],
+        feature_time: Union[str, datetime.datetime],
+        bq_source_uri: str,
+        batch_create_feature_configs: Optional[
```

Comment:
I thought we decide not to do this yet?

And if we do, then it's probably better to use a boolean arg, e.g create_if_missing, which, if turned on, we will compare the existing schema vs the columns in the data source and create features accordingly. It will be filled with some default description, with the name and type best match with data type (there could be cases we map to int but user want double etc, so may not be 100% robust). 

Another option is adding API to infer schema from the data source. It works the same as above, but only return the feature creation request (or feature configs) so that user can patch (e.g update the description, labels) and call batch_features to complete.


Anyway, both these are really fancy and beyond the core for the first delivery. Wdyt?

Reply:
My impression of what we decided not to do (through the design meeting) is not allow users to add, delete, or update/maintain feature configs before sending the batch create feature request. The goal is to ensure stateless within the SDK. But we were discussing utilizing python's native data structure i.e. dictionary, so user can maintain their creating feature id as key without duplicating cases, and SDK can take cares of certain validations, i.e. correct format of `feature_id` (no reserved keywords) and correct type of `value_type`, and `labels`. Also `labels` and `description` are all default to None. We try to avoid map any `type` in general but force user to define, for exactly the reason you mentioned.

For the `batch_create_feature_configs`, users need to specify correctly the `feature_id`. And, no matter what type is used in the BigQuery, users need to specify the `value_type` for the feature in Feature Store.

The implementation for validation and generate the `requests` for `batch_create_features` are in two places:
`entity_type.EntityType._validate_and_get_create_feature_requests` and `featurestore_utils._FeatureConfig.get_create_feature_request`.

May be it's easier to have a call if I misunderstand your question. 

Apart from that. I think having an API to print all features' information is very useful for users. We have so far `list_features` but users need to retrieve those feature level info by iterate through all the listed Feature classes. It will be much simpler at EntityType level to have a glance of all features. However, I would prefer to implement this after the initial scope is implemented. What do you think? Also tagging @sasha-gitg

Reply:
sounds good

Reply:
+1. I think it's cleaner to separate ingest and creation for now, until we gather more feedback.